sky/serve/core.py

```diff
@@ -122,6 +126,14 @@ def up(
     controller_utils.maybe_translate_local_file_mounts_and_sync_up(task,
                                                                    path='serve')
 
+    # If the controller and replicas are from the same cloud, it should
+    # provide better connectivity. We will let the controller choose from
+    # the clouds of the resources if the controller does not exist.
+    requested_clouds: Set['clouds.Cloud'] = set()
+    for resources in task.resources:
+        if resources.cloud is not None:
+            requested_clouds.add(resources.cloud)
```

Comment:
`resources.cloud` is an object and will not be able to be distinguished by `set`.